File: app/src/test/java/openfoodfacts/github/scrachx/openfood/models/entities/country/CountryTest.java
package openfoodfacts.github.scrachx.openfood.models.entities.country;

import org.greenrobot.greendao.DaoException;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.Mock;
import org.mockito.junit.MockitoJUnitRunner;

import java.util.Arrays;
import java.util.List;

import openfoodfacts.github.scrachx.openfood.models.DaoSession;

import static com.google.common.truth.Truth.assertThat;
import static openfoodfacts.github.scrachx.openfood.models.LanguageCodeTestData.LANGUAGE_CODE_ENGLISH;
import static openfoodfacts.github.scrachx.openfood.models.LanguageCodeTestData.LANGUAGE_CODE_FRENCH;
import static openfoodfacts.github.scrachx.openfood.models.entities.country.CountryNameTestData.GERMANY_EN;
import static openfoodfacts.github.scrachx.openfood.models.entities.country.CountryNameTestData.GERMANY_FR;
import static org.junit.Assert.*;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

/**
 * Tests for {@link Country}
 */
@RunWith(MockitoJUnitRunner.class)
public class CountryTest {
    private final String COUNTRY_TAG = "code";
    private final CountryName GERMANY_IN_ENGLISH =
        new CountryName(COUNTRY_TAG, LANGUAGE_CODE_ENGLISH, GERMANY_EN);
    private final CountryName GERMANY_IN_FRENCH =
        new CountryName(COUNTRY_TAG, LANGUAGE_CODE_FRENCH, GERMANY_FR);
    @Mock
    private DaoSession mockDaoSession;
    @Mock
    private CountryDao mockCountryDao;
    @Mock
    private CountryNameDao mockCountryNameDao;
    private Country mCountry;

    @Before
    public void setup() {
        when(mockDaoSession.getCountryDao()).thenReturn(mockCountryDao);
        when(mockDaoSession.getCountryNameDao()).thenReturn(mockCountryNameDao);
        when(mockCountryNameDao._queryCountry_Names(any()))
            .thenReturn(Arrays.asList(GERMANY_IN_ENGLISH, GERMANY_IN_FRENCH));
        mCountry = new Country();
    }

    @Test
    public void getNamesWithNullNamesAndNullDaoSession_throwsDaoException() {
        assertThrows(DaoException.class, () -> mCountry.getNames());
    }

    @Test
    public void getNamesWithNullNamesAndNonNullDaoSession_getsNamesFromCountryNamesDao() {
        mCountry.__setDaoSession(mockDaoSession);

        List<CountryName> names = mCountry.getNames();

        assertThat(names).hasSize(2);
        assertThat(names.get(0).getCountyTag()).isEqualTo(COUNTRY_TAG);
        assertThat(names.get(0).getLanguageCode()).isEqualTo(LANGUAGE_CODE_ENGLISH);
        assertThat(names.get(0).getName()).isEqualTo(GERMANY_EN);

        assertThat(names.get(1).getCountyTag()).isEqualTo(COUNTRY_TAG);
        assertThat(names.get(1).getLanguageCode()).isEqualTo(LANGUAGE_CODE_FRENCH);
        assertThat(names.get(1).getName()).isEqualTo(GERMANY_FR);
    }

    @Test
    public void deleteWithNullDaoSession_throwsDaoException() {
        assertThrows(DaoException.class, () -> mCountry.delete());
    }

    @Test
    public void deleteWithNonNullDaoSession_callsDeleteOnCountryDao() {
        mCountry.__setDaoSession(mockDaoSession);
        mCountry.delete();
        verify(mockCountryDao).delete(mCountry);
    }

    @Test
    public void refreshWithNullDaoSession_throwsDaoException() {
        assertThrows(DaoException.class, () -> mCountry.refresh());
    }

    @Test
    public void refreshWithNonNullDaoSession_callsRefreshOnCountryDao() {
        mCountry.__setDaoSession(mockDaoSession);
        mCountry.refresh();
        verify(mockCountryDao).refresh(mCountry);
    }

    @Test
    public void updateWithNullDaoSession_throwsDaoException() {
        assertThrows(DaoException.class, () -> mCountry.update());
    }

    @Test
    public void updateWithNonNullDaoSession_callsUpdateOnCountryDao() {
        mCountry.__setDaoSession(mockDaoSession);
        mCountry.update();
        verify(mockCountryDao).update(mCountry);
    }
}

File: app/src/test/java/openfoodfacts/github/scrachx/openfood/models/entities/country/CountryTest.kt
package openfoodfacts.github.scrachx.openfood.models.entities.country

import com.google.common.truth.Truth.assertThat
import openfoodfacts.github.scrachx.openfood.models.DaoSession
import openfoodfacts.github.scrachx.openfood.models.LanguageCodeTestData.LANGUAGE_CODE_ENGLISH
import openfoodfacts.github.scrachx.openfood.models.LanguageCodeTestData.LANGUAGE_CODE_FRENCH
import openfoodfacts.github.scrachx.openfood.models.entities.country.CountryNameTestData.GERMANY_EN
import openfoodfacts.github.scrachx.openfood.models.entities.country.CountryNameTestData.GERMANY_FR
import org.greenrobot.greendao.DaoException
import org.junit.Assert.assertThrows
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import org.mockito.Mock
import org.mockito.junit.MockitoJUnitRunner
import org.mockito.kotlin.*

/**
 * Tests for [Country]
 */
@RunWith(MockitoJUnitRunner::class)
class CountryTest {

    @Mock
    private lateinit var mockDaoSession: DaoSession

    @Mock
    private lateinit var mockCountryDao: CountryDao

    @Mock
    private lateinit var mockCountryNameDao: CountryNameDao
    private lateinit var country: Country

    @Before
    fun setup() {
        whenever(mockDaoSession.countryDao) doReturn mockCountryDao
        whenever(mockDaoSession.countryNameDao) doReturn mockCountryNameDao
        whenever(mockCountryNameDao._queryCountry_Names(anyOrNull())) doReturn listOf(GERMANY_IN_ENGLISH, GERMANY_IN_FRENCH)

        country = Country()
    }

    @Test
    fun `getting names with null names and null dao session throws DaoException`() {
        assertThrows(DaoException::class.java) { country.names }
    }

    @Test
    fun `getting names with null names`() {
        country.__setDaoSession(mockDaoSession)

        val names = country.names
        assertThat(names).hasSize(2)

        val countryName1 = names[0]!!
        assertThat(countryName1.countyTag).isEqualTo(COUNTRY_TAG)
        assertThat(countryName1.languageCode).isEqualTo(LANGUAGE_CODE_ENGLISH)
        assertThat(countryName1.name).isEqualTo(GERMANY_EN)

        val countryName2 = names[1]!!
        assertThat(countryName2.countyTag).isEqualTo(COUNTRY_TAG)
        assertThat(countryName2.languageCode).isEqualTo(LANGUAGE_CODE_FRENCH)
        assertThat(countryName2.name).isEqualTo(GERMANY_FR)
    }

    @Test
    fun `delete with null daoSession throws DaoException`() {
        assertThrows(DaoException::class.java) { country.delete() }
    }

    @Test
    fun `delete calls delete on CountryDao`() {
        country.__setDaoSession(mockDaoSession)
        country.delete()

        verify(mockCountryDao).delete(country)
    }

    @Test
    fun `refresh with null daoSession throws DaoException`() {
        assertThrows(DaoException::class.java) { country.refresh() }
    }

    @Test
    fun `refresh calls refresh on CountryDao`() {
        country.__setDaoSession(mockDaoSession)
        country.refresh()
        verify(mockCountryDao).refresh(country)
    }

    @Test
    fun `update with null daoSession should throw DaoException`() {
        assertThrows(DaoException::class.java) { country.update() }
    }

    @Test
    fun `update calls update on CountryDao`() {
        country.__setDaoSession(mockDaoSession)
        country.update()
        verify(mockCountryDao).update(country)
    }

    companion object {
        private const val COUNTRY_TAG = "code"
        private val GERMANY_IN_ENGLISH = CountryName(COUNTRY_TAG, LANGUAGE_CODE_ENGLISH, GERMANY_EN)
        private val GERMANY_IN_FRENCH = CountryName(COUNTRY_TAG, LANGUAGE_CODE_FRENCH, GERMANY_FR)
    }
}